lib/awesome_spawn.rb

```diff
@@ -116,15 +138,94 @@ def build_command_line(command, params = nil)
+    end
     return output, error, status && status.exitstatus
   end
 
-  def parse_command_options(command, options)
+  def normalize_run_opts(*opts)
+    commands, options = if opts.last.kind_of?(Hash)
+                          [opts[0..-2], opts.last]
+                        else
+                          [opts, {}]
+                        end
+
+    if commands.all? { |cmd| cmd.kind_of?(Array) }
+      commands = commands.flatten(1)
+    end
+
+    bad_keys = (options.keys.flatten & [:in, :out, :err]).map { |k| ":#{k}" }
+    raise ArgumentError, "options cannot contain #{bad_keys.join(", ")}" if bad_keys.any?
+
+    [Array(commands), options]
+  end
+
+  def parse_command_options(commands, options)
     options = options.dup
     params  = options.delete(:params)
     env = options.delete(:env) || {}
 
-    [env, build_command_line(command, params), options]
+    cmds = if commands.length == 1
+             build_command_line(commands.first, params)
+           else
+             commands.map do |(cmd, args)|
+               build_command_line(cmd, (args || {})[:params])
+             end
+           end
+
+    [env, cmds, options]
+  end
+
+  # This is a combination of the functionality of Open3.capture3 and
+  # Open3.pipeline_r
+  #
+  # Open3.pipeline_r doesn't support capturing stderr, so this method takes the
+  # setup code found in `Open3#popen3`, and the block from `Open3#capture3`,
+  # and combines them into a single method.
+  #
+  # Should have the same interface as `Open3.capture3`, but supports passing in
+  # pipes.
+  def pipe_capture3(env, cmds, stdin_data: '', binmode: false, **opts)
+    # Begin excerpt from Open3.popen3
+    in_r, in_w = IO.pipe
+    opts[:in] = in_r
+    in_w.sync = true
+
+    out_r, out_w = IO.pipe
+    opts[:out] = out_w
+
+    err_r, err_w = IO.pipe
+    opts[:err] = err_w
+    # End excerpt from Open3.popen3
+
+    # Inject global ENV to all commands
+    commands   = cmds.map { |cmd| Array(cmd).unshift(env) }
+    child_ios  = [in_r, out_w, err_w]
+    parent_ios = [in_w, out_r, err_r]
+
+    pipeline_run(commands, opts, child_ios, parent_ios) do |*result|
+      # Mostly the same as the block content in Open3#capture3, except that
+      # pipeline_run returns an array of Processes that have been piped to, and
+      # we just want the status of the last one.
+      pipe_in, pipe_out, pipe_err, child_pids = result
+      # Begin block excerpt from Open3#capture3
+      if binmode
+        pipe_in.binmode
+        pipe_out.binmode
+        pipe_err.binmode
+      end
+      out_reader = Thread.new { pipe_out.read }
+      err_reader = Thread.new { pipe_err.read }
+      begin
+        pipe_in.write(stdin_data)
+      rescue Errno::EPIPE # rubocop:disable Lint/HandleExceptions (original code)
+      end
+      pipe_in.close
+      # End block excerpt from Open3#capture3
+      [out_reader.value, err_reader.value, child_pids.last.value]
+    end
```

Comment:
Cool...I wonder if you can contribute back to Ruby core.  This seems really useful.